File: backend/app/agents/subtitle_director.py
from pathlib import Path
from typing import Dict, Any, List
from app.agents.base import BaseAgent

def format_ts(seconds: float, sep: str = ",") -> str:
    hrs = int(seconds // 3600)
    mins = int((seconds % 3600) // 60)
    secs = int(seconds % 60)
    ms = int((seconds % 1.0) * 1000)
    return f"{hrs:02d}:{mins:02d}:{secs:02d}{sep}{ms:03d}"

class SubtitleDirectorAgent(BaseAgent):
    def __init__(self, base_storage_dir: str = "./storage/runs"):
        super().__init__("subtitle_director")
        self.base_storage_dir = Path(base_storage_dir)

    async def _execute(self, job_id: str, scene_id: str, context: Dict[str, Any], retry_count: int) -> Dict[str, Any]:
        target_lang = context.get("target_language", "en").lower().split("-")[0]
        localized_lines = context.get("localized_lines", [])
        sync_adjustments = context.get("sync_adjustments", [])

        sync_map = {adj["segment_id"]: float(adj["final_duration_s"]) for adj in sync_adjustments}

        cues = []
        max_cps = 0.0
        drift_detected = False

        for idx, line in enumerate(localized_lines):
            seg_id = line.get("segment_id", idx + 1)
            start_s = float(line.get("start_s", 0.0))
            duration_s = sync_map.get(seg_id, float(line.get("end_s", 1.0) - start_s))
            end_s = round(start_s + duration_s, 3)
            text = line.get("translated_text", "").strip()

            cps = round(len(text) / duration_s, 1) if duration_s > 0 else 0.0
            if cps > max_cps:
                max_cps = cps

            # Check drift against original end
            orig_end = float(line.get("end_s", end_s))
            if abs(end_s - orig_end) > 0.5:
                drift_detected = True

            cues.append({
                "cue_index": idx + 1,
                "start_s": start_s,
                "end_s": end_s,
                "text": text,
                "cps": cps
            })

        # Generate SRT
        srt_lines = []
        for c in cues:
            srt_lines.append(f"{c['cue_index']}\n{format_ts(c['start_s'], ',')} --> {format_ts(c['end_s'], ',')}\n{c['text']}\n")
        srt_content = "\n".join(srt_lines)

        # Generate VTT
        vtt_lines = ["WEBVTT\n"]
        for c in cues:
            vtt_lines.append(f"{c['cue_index']}\n{format_ts(c['start_s'], '.')} --> {format_ts(c['end_s'], '.')}\n{c['text']}\n")
        vtt_content = "\n".join(vtt_lines)

        run_dir = self.base_storage_dir / job_id
        run_dir.mkdir(parents=True, exist_ok=True)
        srt_path = run_dir / f"subtitles_{target_lang}.srt"
        vtt_path = run_dir / f"subtitles_{target_lang}.vtt"

        with open(srt_path, "w", encoding="utf-8") as f:
            f.write(srt_content)
        with open(vtt_path, "w", encoding="utf-8") as f:
            f.write(vtt_content)

        decision = (
            f"Subtitle Director formatted {len(cues)} cues into SRT and VTT for {target_lang.upper()}, "
            f"max CPS: {max_cps}, drift_detected: {drift_detected}."
        )

        return {
            "job_id": job_id,
            "target_language": target_lang,
            "srt_path": str(srt_path),
            "vtt_path": str(vtt_path),
            "total_cues": len(cues),
            "max_cps": max_cps,
            "drift_detected": drift_detected,
            "decision": decision,
            "quality_score": 95.0 if not drift_detected else 82.0
        }

File: frontend/lib/agents/subtitle_director.ts
export interface SubtitleCue {
  cue_index: number;
  segment_id: number;
  speaker_id: string;
  start_s: number;
  end_s: number;
  duration_s: number;
  text: string;
  cps: number;
  drift_detected: boolean;
}

export interface SubtitleDirectorInput {
  job_id: string;
  target_language: string;
  localized_lines: Array<{
    segment_id: number;
    speaker_id: string;
    start_s: number;
    end_s: number;
    translated_text: string;
  }>;
  sync_adjustments?: Array<{
    segment_id: number;
    final_duration_s: number;
  }>;
}

export interface SubtitleDirectorOutput {
  job_id: string;
  target_language: string;
  srt_path: string;
  vtt_path: string;
  total_cues: number;
  max_cps: number;
  drift_detected: boolean;
}

function formatTimestamp(seconds: number, separator: ',' | '.'): string {
  const hrs = Math.floor(seconds / 3600);
  const mins = Math.floor((seconds % 3600) / 60);
  const secs = Math.floor(seconds % 60);
  const ms = Math.floor((seconds % 1) * 1000);

  const hh = String(hrs).padStart(2, '0');
  const mm = String(mins).padStart(2, '0');
  const ss = String(secs).padStart(2, '0');
  const mmm = String(ms).padStart(3, '0');

  return `${hh}:${mm}:${ss}${separator}${mmm}`;
}

export function generateSubtitleCues(
  localized_lines: SubtitleDirectorInput['localized_lines'],
  sync_adjustments?: SubtitleDirectorInput['sync_adjustments']
): SubtitleCue[] {
  const syncMap = new Map<number, number>();
  if (sync_adjustments) {
    for (const adj of sync_adjustments) {
      syncMap.set(adj.segment_id, adj.final_duration_s);
    }
  }

  return localized_lines.map((line, idx) => {
    const adjustedDuration = syncMap.get(line.segment_id) || (line.end_s - line.start_s);
    const start_s = Number(line.start_s.toFixed(3));
    const end_s = Number((start_s + adjustedDuration).toFixed(3));
    const duration_s = Number(adjustedDuration.toFixed(3));
    const text = line.translated_text.trim();
    const cps = duration_s > 0 ? Number((text.length / duration_s).toFixed(1)) : 0;

    return {
      cue_index: idx + 1,
      segment_id: line.segment_id,
      speaker_id: line.speaker_id,
      start_s,
      end_s,
      duration_s,
      text,
      cps,
      drift_detected: false,
    };
  });
}

export function formatSrt(cues: SubtitleCue[]): string {
  return cues
    .map(
      (cue) =>
        `${cue.cue_index}\n${formatTimestamp(cue.start_s, ',')} --> ${formatTimestamp(cue.end_s, ',')}\n${cue.text}\n`
    )
    .join('\n');
}

export function formatVtt(cues: SubtitleCue[]): string {
  const body = cues
    .map(
      (cue) =>
        `${cue.cue_index}\n${formatTimestamp(cue.start_s, '.')} --> ${formatTimestamp(cue.end_s, '.')}\n${cue.text}\n`
    )
    .join('\n');
  return `WEBVTT\n\n${body}`;
}

export function validateSubtitleDirectorOutput(output: any): output is SubtitleDirectorOutput {
  if (!output || typeof output !== 'object') return false;
  if (typeof output.job_id !== 'string') return false;
  if (typeof output.target_language !== 'string') return false;
  if (typeof output.srt_path !== 'string') return false;
  if (typeof output.vtt_path !== 'string') return false;
  if (typeof output.total_cues !== 'number') return false;
  if (typeof output.max_cps !== 'number') return false;
  if (typeof output.drift_detected !== 'boolean') return false;
  return true;
}
